In [1]:
import pandas as pd
import numpy as np
import warnings
warnings.filterwarnings("ignore")

df = pd.read_csv("carsclean.csv")
df.head()

city car_maker     model  ...  color            pay_method    price
0  Al Riyadh    Toyota      Rush  ...  White  Cash or Installments  58375.0
1  Al Riyadh       Kia  Carnival  ...   Grey             Cash Only  10000.0
2  Al Riyadh    Toyota     Camry  ...  White     Installments Only  58375.0
3  Al Riyadh      Ford    Fusion  ...  White             Cash Only  22425.4
4  Al Riyadh      Ford    Taurus  ...   Grey             Cash Only  22425.4

[5 rows x 11 columns]

In [2]:
df.info()   #Hiç bir sütunda null değeri olmadığını görüyoruz.

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 560 entries, 0 to 559
Data columns (total 11 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   city          560 non-null    object 
 1   car_maker     560 non-null    object 
 2   model         560 non-null    object 
 3   year          560 non-null    int64  
 4   condition     560 non-null    object 
 5   kilometers    560 non-null    float64
 6   transmission  560 non-null    object 
 7   fuel          560 non-null    object 
 8   color         560 non-null    object 
 9   pay_method    560 non-null    object 
 10  price         560 non-null    float64
dtypes: float64(2), int64(1), object(8)
memory usage: 48.2+ KB


In [3]:
columns_list = df.columns  #Sütun isimlerini columns_list değişkeninde depoladık.
                           #Listenin her bir ögesinde dolaşarak sütun isimlerini alıp
                           #O sütunda kaç çeşit değişken olduğunu yazdırdık
for column in columns_list:
  print(column)
  print(df[column].value_counts())
  print("--------------------")

city
Al Riyadh    560
Name: city, dtype: int64
--------------------
car_maker
Toyota      140
Ford        120
BMW          80
Kia          60
Renault      40
Nissan       20
Cadillac     20
Mazda        20
Suzuki       20
Dodge        20
GMC          20
Name: car_maker, dtype: int64
--------------------
model
Camry             40
740               40
Fusion            40
Other             20
Rio               20
Rush              20
Crown Victoria    20
Symbol            20
Avalon            20
X5                20
Patrol            20
Opirus            20
Dokker Van        20
Carnival          20
Explorer          20
Yaris             20
Durango           20
3                 20
Prado             20
Fortuner          20
Escalade          20
Expedition        20
Envoy             20
730               20
Taurus            20
Name: model, dtype: int64
--------------------
year
2008    100
2014     80
2016     60
2013     60
2007     60
2019     40
2009     40
2018     20
2015     20
2012

In [0]:
df.drop(labels=['city','fuel'],axis=1,inplace=True)  #City ve Fuel sütunları tek değişken içerdiği için modele katkısı olmucaktır.
                                                     #Bu yüzden bu sütunları siliyoruz. axis=0 row , axis=1 sütun demektir.
                                                     #inplace=True verinin üzerinde değişiklik yapar. False olsaydı yaptığımız şey kaydolmucaktı.

In [5]:
df.head()  #yeni halini görelim

car_maker     model  year  ...  color            pay_method    price
0    Toyota      Rush  2019  ...  White  Cash or Installments  58375.0
1       Kia  Carnival  2006  ...   Grey             Cash Only  10000.0
2    Toyota     Camry  2014  ...  White     Installments Only  58375.0
3      Ford    Fusion  2010  ...  White             Cash Only  22425.4
4      Ford    Taurus  2018  ...   Grey             Cash Only  22425.4

[5 rows x 9 columns]

In [0]:
#get_dummies metoduyla her birinin içindeki değerleri çıkartıp kendi ismiyle bir değişkene atadık.
df_car_maker = pd.get_dummies(df.car_maker)
df_model = pd.get_dummies(df.model)
df_condition = pd.get_dummies(df.condition)
df_transmission = pd.get_dummies(df.transmission)
df_color = pd.get_dummies(df.color)
df_pay_method = pd.get_dummies(df.pay_method)

In [8]:
#Bunlardan bir tanesi neymiş görelim
df_pay_method

Cash Only  Cash or Installments  Installments Only
0            0                     1                  0
1            1                     0                  0
2            0                     0                  1
3            1                     0                  0
4            1                     0                  0
..         ...                   ...                ...
555          1                     0                  0
556          1                     0                  0
557          1                     0                  0
558          1                     0                  0
559          1                     0                  0

[560 rows x 3 columns]

In [0]:
#Tüm kategorik sütunları kendilerine has bir değişkende depoladığımıza göre aynı metodla bunlarıda df'mizden silelim.
df.drop(labels=['car_maker','model','condition','transmission','color','pay_method'],axis=1,inplace=True)

In [10]:
df.head()  #son halini görelim

year  kilometers    price
0  2019     14999.5  58375.0
1  2006         0.0  10000.0
2  2014     24999.5  58375.0
3  2010    200000.0  22425.4
4  2018      5000.0  22425.4

In [0]:
#Kalan 3 sütunluk verimizle , değişkenlere atadığımız diğer sütunları birleştirelim
#pandas.concat birleştirmek istediğimiz df'leri liste içinde kabul eder.
#axis=1 yine sütun olarak yani yanyana birleştir anlamına gelir.
dataframe = pd.concat([df,df_car_maker,df_model,df_condition,df_transmission,df_color,df_pay_method],axis=1)

In [12]:
dataframe.head()   #Evet başarıyla yanyana birleştirdik.

year  kilometers  ...  Cash or Installments  Installments Only
0  2019     14999.5  ...                     1                  0
1  2006         0.0  ...                     0                  0
2  2014     24999.5  ...                     0                  1
3  2010    200000.0  ...                     0                  0
4  2018      5000.0  ...                     0                  0

[5 rows x 53 columns]

In [13]:
#Şimdi biraz ellerimizi kirletelim. Price tahmini yapacağımız için price'ın en son sütunda olmasını istiyorum.
#Öncelikle sütun isimlerimizi çekelim ve liste halinde depolayalım
dataframe_columns_list = list(dataframe.columns.values)
dataframe_columns_list

['year',
 'kilometers',
 'price',
 'BMW',
 'Cadillac',
 'Dodge',
 'Ford',
 'GMC',
 'Kia',
 'Mazda',
 'Nissan',
 'Renault',
 'Suzuki',
 'Toyota',
 '3',
 '730',
 '740',
 'Avalon',
 'Camry',
 'Carnival',
 'Crown Victoria',
 'Dokker Van',
 'Durango',
 'Envoy',
 'Escalade',
 'Expedition',
 'Explorer',
 'Fortuner',
 'Fusion',
 'Opirus',
 'Other',
 'Patrol',
 'Prado',
 'Rio',
 'Rush',
 'Symbol',
 'Taurus',
 'X5',
 'Yaris',
 'New',
 'Used',
 'Automatic',
 'Manual',
 'Beige',
 'Black',
 'Brown',
 'Green',
 'Grey',
 'Silver',
 'White',
 'Cash Only',
 'Cash or Installments',
 'Installments Only']

In [14]:
dataframe_columns_list.remove('price')   #price değerini listeden sildik.
dataframe_columns_list.append('price')   #append metoduyla 'price' değerini listenin en sonuna geri koyduk.
dataframe_columns_list                   #evet price en sona başarıyla gitmiş

['year',
 'kilometers',
 'BMW',
 'Cadillac',
 'Dodge',
 'Ford',
 'GMC',
 'Kia',
 'Mazda',
 'Nissan',
 'Renault',
 'Suzuki',
 'Toyota',
 '3',
 '730',
 '740',
 'Avalon',
 'Camry',
 'Carnival',
 'Crown Victoria',
 'Dokker Van',
 'Durango',
 'Envoy',
 'Escalade',
 'Expedition',
 'Explorer',
 'Fortuner',
 'Fusion',
 'Opirus',
 'Other',
 'Patrol',
 'Prado',
 'Rio',
 'Rush',
 'Symbol',
 'Taurus',
 'X5',
 'Yaris',
 'New',
 'Used',
 'Automatic',
 'Manual',
 'Beige',
 'Black',
 'Brown',
 'Green',
 'Grey',
 'Silver',
 'White',
 'Cash Only',
 'Cash or Installments',
 'Installments Only',
 'price']

In [15]:
#Bu bilgi stackoverflow'dan alınmıştır. Sütun sırasını değiştirme formulü --> df = df[sütun_listesi]

dataframe = dataframe[dataframe_columns_list]   #Yeni sütun sıralamamızı
dataframe.head()

year  kilometers  BMW  ...  Cash or Installments  Installments Only    price
0  2019     14999.5    0  ...                     1                  0  58375.0
1  2006         0.0    0  ...                     0                  0  10000.0
2  2014     24999.5    0  ...                     0                  1  58375.0
3  2010    200000.0    0  ...                     0                  0  22425.4
4  2018      5000.0    0  ...                     0                  0  22425.4

[5 rows x 53 columns]

In [17]:
#Ve datapreprocessing'imizin son partı :)
corr_matrix = dataframe.corr()  #dataframe.corr() her bir sütun arasındaki korelasyona bakar. Biz bunu corr_matrix değişkenine eşitledik.
corr_matrix['price':]    #corr_matrix['price'] diyerek 'price' yi filtreledik fakat 1D array oluşturduk. 2D olması için sonuna : ekledik.

year  kilometers  ...  Installments Only  price
price  0.255938   -0.077572  ...          -0.074751    1.0

[1 rows x 53 columns]


Price değişkeninin her bir sütunla yaptığın korelasyon değerini görebilirsiniz. 1.00 tam korelasyon , -1.00 negatif yönde tam korelasyondur. 0.00 ise aralarında korelasyon yoktur yani 2 değişkenden birinin belli bir kovaryansı yoktur. BÖLÜM 1 SONU :)

####PART2 - Machine Learning


In [18]:
df = pd.read_csv("carsclean.csv")                     #verimizi import ettik
df.drop(labels=['city','fuel'],axis=1,inplace=True)   #city  ve fuel değişkenleri işimize yaramadığı için yeniden sildik
df.head()  #df'imizin güncel hali

car_maker     model  year  ...  color            pay_method    price
0    Toyota      Rush  2019  ...  White  Cash or Installments  58375.0
1       Kia  Carnival  2006  ...   Grey             Cash Only  10000.0
2    Toyota     Camry  2014  ...  White     Installments Only  58375.0
3      Ford    Fusion  2010  ...  White             Cash Only  22425.4
4      Ford    Taurus  2018  ...   Grey             Cash Only  22425.4

[5 rows x 9 columns]

In [0]:
# df.iloc satır ve sütun seçme komutumuzdur. [:, ] virgülden önceki iki nokta tüm satırları seçmemizi sağlar.
# Virgülden sonrası sütun seçme operasyonudur. X için son sütün hariç tüm satır ve sütünları seçtik.
# y içinde tüm satırları ve son sütunu seçtik. Sonuna .values ekleyerek değerlerimizi numpy-array olarak aldık.
# Numpy-array'e dönüştürmemizin nedeni sklearn kütüphanesinin numpy verileri istemesinden dolayı.
# İsterseniz farkı görmek için "df.iloc[:, :-1]" ve "df.iloc[:, :-1].values" değerlerini ayrı ayrı çalıştırın.
X = df.iloc[:, :-1].values 
y = df.iloc[:, -1:].values

In [20]:
X[:1] #X'in ilk satırına bakalım

array([['Toyota', 'Rush', 2019, 'Used', 14999.5, 'Automatic', 'White',
        'Cash or Installments']], dtype=object)

In [0]:
from sklearn.preprocessing import LabelEncoder   #Label encoder kategorik verileri , nümerik verilere çevirir.
labelencoder = LabelEncoder()    #LabelEncoder'ı labelencoder değişkenine atadık.
#Şimdi her bir kategorik sütunu labelencoder'la encode'licez. Neden mi? Makine yazıları anlayamaz sayıları anlar bu yüzden :)

X[:,0] = labelencoder.fit_transform(X[:,0]) #0. sütun araba markamız olan car_marker
X[:,1] = labelencoder.fit_transform(X[:,1]) #1. sütun modelimiz 
X[:,3] = labelencoder.fit_transform(X[:,3]) #3. sütun condition değerimiz
X[:,5] = labelencoder.fit_transform(X[:,5]) #5. sütun transmission değerimiz yani araç auto mu manuel mi.
X[:,6] = labelencoder.fit_transform(X[:,6]) #6. sütun arabamızın rengi olan color sütunu
X[:,7] = labelencoder.fit_transform(X[:,7]) #7. sütun ise ödeme metodu

Yukarıdaki koda alternatif olarak böylede yapabilirdik

```
categorical_columns = [0,1,3,5,6,7]

def categorical_to_numerical(column_number):
  X[:,column_number] = labelencoder.fit_transform(X[:,column_number])   

for i in range(len(categorical_columns)):
  categorical_to_numerical(categorical_columns[i])
```

Seçim size kalmış aynı şey :)

In [22]:
#Ödeme metodumuzun ilk 5 satırı sırasıyla : Cash or Installments,Cash Only,Installments Only,Cash Only,Cash Only	'idi.

X[:5,7:]  #Şimdi ödeme metodumuzun ilk 5 satırını neye dönüştürdüğümüzü görelim.

array([[1],
       [0],
       [2],
       [0],
       [0]], dtype=object)

Sklearn alfabetik sıralamaya göre numaralandırır. Bu yüzden örnek olarak

* Cash Only : 0
* Cash or Installments : 1
* Installments Only : 2 'dir. 

Bir sonraki aşamada OneHotEncoding yapıcaz :)

to_catogorical() metoduyla OneHotEncoding yapmak istediğimiz sütunu seçicez. Böylece 1 sütunumuz içindeki değişken veri sayısı kadar büyücek.

Az önceki örnek için;

Cash Only || Cash or Installments || Installments Only

* 1 || 0 || 0
* 0 || 1 || 0
* 0 || 0 || 1

Olucak. Gölge değişken tuzağından kaçınmamız için 1 sütunu silmemiz gerek. Installment's Only'e ihtiyacımız yok çünkü ilk 2 sütun "0" olduğunda otomatikmen Installments Only 1 anlamına gelir. O yüzden sonlarına [:,:-1] ekleyerek oluşturduğumuz OneHotEncodinglerin son sütununu dahil etmicez.

Ve son olarak eğer kategorik değişkenimiz sadece 2 eşsiz değerden oluşuyorsa OneHotEncoding yapmamıza gerek yok çünkü sütun değerleri zaten 0 veya 1 dir.

In [23]:
from keras.utils import to_categorical #Bir derin öğrenme kütüphanesi olan keras'ın nimetlerinden faydalanalım.

car_maker = to_categorical(X[:,0])[:,:-1]   #2'den fazla kategori bulundurduğu için OneHotEncoding yaptık.
model = to_categorical(X[:,1])[:,:-1]       
year = X[:,2:3].astype('float32')           #Year değişkenini 2. sütundan 3'e kadar seçtik 3 dahil değil.
condition =  X[:,3:4].astype('float32')     #Sonuna float32 ekleyerek data türümüzü garantileyelim
kilometers = X[:,4:5].astype('float32') 
transmission = X[:,5:6].astype('float32')
color = to_categorical(X[:,6])[:,:-1]       
pay_method = to_categorical(X[:,7])[:,:-1]

Using TensorFlow backend.


Bakalım ne el ettik ilk 5 satırı hepsinde görünteleyelim

In [24]:
def ne_elde_ettik(sutun,name):
  print(name)
  print(sutun[:3] , sutun.shape)
  print("------------------------------")

xd = [car_maker,model,year,condition,kilometers,transmission,color,pay_method]
xdxd = ["car_maker","model","year","condition","kilometers","transmission","color","pay_method"]

for i in range(len(xd)):
  ne_elde_ettik(xd[i],xdxd[i])

car_maker
[[0. 0. 0. 0. 0. 0. 0. 0. 0. 0.]
 [0. 0. 0. 0. 0. 1. 0. 0. 0. 0.]
 [0. 0. 0. 0. 0. 0. 0. 0. 0. 0.]] (560, 10)
------------------------------
model
[[0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 1. 0. 0. 0.]
 [0. 0. 0. 0. 0. 1. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0.]
 [0. 0. 0. 0. 1. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0.]] (560, 24)
------------------------------
year
[[2019.]
 [2006.]
 [2014.]] (560, 1)
------------------------------
condition
[[1.]
 [1.]
 [1.]] (560, 1)
------------------------------
kilometers
[[14999.5]
 [    0. ]
 [24999.5]] (560, 1)
------------------------------
transmission
[[0.]
 [0.]
 [0.]] (560, 1)
------------------------------
color
[[0. 0. 0. 0. 0. 0.]
 [0. 0. 0. 0. 1. 0.]
 [0. 0. 0. 0. 0. 0.]] (560, 6)
------------------------------
pay_method
[[0. 1.]
 [1. 0.]
 [0. 0.]] (560, 2)
------------------------------


In [0]:
#Numpy'ın concantenate metoduyla tüm oluşturmuş olduğumuz(En son sütunu çıkarılmış sütünları) yanyana ekliyoruz. axis=1 sütun olarak birleştir
X = np.concatenate((car_maker,model,year,condition,kilometers,transmission,color,pay_method),axis=1) #Yeni X'imiz.

In [26]:
print(X , X.shape)  #560 satır 46 sütunluk bomba gibi verimiz hazır.

[[0. 0. 0. ... 0. 0. 1.]
 [0. 0. 0. ... 0. 1. 0.]
 [0. 0. 0. ... 0. 0. 0.]
 ...
 [1. 0. 0. ... 0. 1. 0.]
 [0. 0. 0. ... 0. 1. 0.]
 [0. 0. 0. ... 0. 1. 0.]] (560, 46)


Bir sonraki bölümde makine öğrenmesi kısmına tam olarak başlıyoruz. En zevkli part :)

X bağımsız değişkenlerimiz ve y bağımlı değişkenimizi train ve test datası olarak 2 ye ayıralım. Test datamız ana verimizin %20 sini içericek. Sonuçların aynı çıkması için random_state=0 olarak kalsın.

In [0]:
from sklearn.model_selection import train_test_split   #train,test ayırmak için gerekli kütüphane
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.2, random_state = 0)

In [28]:
from sklearn.linear_model import LinearRegression    #Tekli ve Çoklu regresyon için kullandığımız kütüphane
regressor = LinearRegression()                       #LinearRegression() metodumuzu regressor değişkenine atadık.
regressor.fit(X_train, y_train)                      #.fit metoduyla X_train ve y_train'imizi modelimize uydurduk ve öğrettik.

LinearRegression(copy_X=True, fit_intercept=True, n_jobs=None, normalize=False)

Evet Bu kadar. 2 satır kod ile makinemiz öğrenmiş oldu. Şimdi gelelim sonuçlara :)

In [0]:
#.predict metodunun içine X_test'i vererek tahmin etmesini sağladık. Ve tahminlerimizi y_pred'e atadık.
y_pred = regressor.predict(X_test)

In [30]:
from sklearn.metrics import r2_score  #r2 score'mizi test edelim.
r2_score(y_test, y_pred)

0.9998422923910831

Mükemmel! Tüm tahminlerimizi ve gerçek değerleri bir dataframe üzerinde görüntüleyelim

In [31]:
tahmin = np.concatenate((y_pred,y_test),axis=1)
pd.DataFrame(tahmin , columns=["Tahminimiz","Gerçek Değerler"])

Tahminimiz  Gerçek Değerler
0       58379.5          58375.0
1       41633.5          42000.0
2       24981.5          25000.0
3       58678.0          58375.0
4        1759.0           2114.0
..          ...              ...
107     98358.0          98500.0
108     21876.0          22000.0
109     22823.0          23000.0
110     45077.0          45000.0
111     58678.0          58375.0

[112 rows x 2 columns]

In [34]:
regressor.intercept_ , regressor.coef_  #bias ve coef değerlerimiz. Bu kadar yüksek çıkmasının sebebi verimize herhangi bir standardazation veya normalization uygulamamış olmamızdan

(array([-5074691.5], dtype=float32),
 array([[ 2.7089192e+03, -3.1715579e+02,  6.4352130e+02, -1.8260869e+04,
         -5.5932192e+03, -1.6271732e+04, -3.8530190e+03,  4.2957139e+03,
         -2.4141479e+04, -9.5425156e+03, -3.8530300e+03,  2.2121988e+04,
          1.7496949e+04, -5.5673354e+03,  1.7130414e+04,  2.3481414e+03,
          2.2511965e+03, -5.5956318e+03,  6.4354120e+02, -5.5932109e+03,
         -3.8731888e+02, -2.5297773e+04,  1.7470033e+04,  3.4548883e+04,
          5.3037573e+02, -2.5729150e+03, -9.5425215e+03,  4.2957109e+03,
          2.2847148e+04, -1.6046967e+04, -6.0739292e+03, -1.8545844e+04,
         -1.3214695e+04, -7.2574463e+03,  2.5488152e+03, -1.6322127e+04,
         -3.0120850e-02, -1.0842438e+04, -1.6046966e+04,  9.0416582e+03,
          2.2121990e+04, -2.5988822e+04, -1.2795910e+04, -6.0820625e+03,
          1.4628604e+04,  9.8608789e+03]], dtype=float32))

Fiyat tahmini ve gerçek fiyat değerleri oldukça yakın. Şimdi spesifik bir satıra bakalım

X_test'imizde 50. satırımızın

* Markası : Mazda
* Modeli : 3
* Yılı : 2013
* Durumu : 2.el
* Kilometresi : 154.999
* Vites : Otomatik
* Rengi : Yeşil
* Ödeme tipi : Nakit
* Fiyatı : 16000

In [32]:
y_pred = regressor.predict(X_test[50:51,:])
y_pred

array([[16016.5]], dtype=float32)

Yapay zekamız fiyatı 16016 olarak tahmin etti. Kodları hep birlikte geliştirme dileğiyle sevgiler :)